### PR TITLE
Linux 4.18 compat: inode timespec -> timespec64

### PR DESCRIPTION
**Pull request by behlendorf**
### Description

Commit torvalds/linux@95582b0 changes the inode i_atime, i_mtime, and i_ctime members form timespec's to timespec64's to make them 2038 safe.  As part of this change the current_time() function was also updated to return the timespec64 type.
    
Resolve this issue by introducing a new inode_timespec_t type which is defined to match the timespec type used by the inode.  It should be used when working with inode timestamps to ensure matching types.
    
The timestruc_t type under Illumos was used in a similar fashion but was specified to always be a timespec_t.  Rather than incorrectly define this type all timespec_t types have been replaced by the new inode_timespec_t type.
    
Finally, the kernel and user space 'sys/time.h' headers were aligned with each other.  They define as appropriate for the context several constants as macros and include static inline implementation of gethrestime(), gethrestime_sec(), and gethrtime().
    
### Motivation and Context

Support the latest kernels.

### How Has This Been Tested?

Locally built against v4.18-rc1-23-g9ffc59d57228.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: module/zfs/zpl_ctldir.c [hunk 1]
```diff
@@ -388,7 +388,8 @@ zpl_snapdir_getattr_impl(const struct path *path, struct kstat *stat,
 	generic_fillattr(ip, stat);
 
 	stat->nlink = stat->size = 2;
-	stat->ctime = stat->mtime = dmu_objset_snap_cmtime(zfsvfs->z_os);
+	stat->ctime = stat->mtime = zpl_inode_set_timespec(
+	    dmu_objset_snap_cmtime(zfsvfs->z_os));
```
Comment (tuxoko) [inline module/zfs/zpl_ctldir.c]: Can we make this timespec64 as well?

File: module/zfs/zfs_znode.c [hunk 2]
```diff
@@ -710,8 +710,9 @@ zfs_mknode(znode_t *dzp, vattr_t *vap, dmu_tx_t *tx, cred_t *cr,
 	znode_hold_t	*zh;
 
 	if (zfsvfs->z_replay) {
+		/* see zfs_replay_create() */
 		obj = vap->va_nodeid;
-		now = vap->va_ctime;		/* see zfs_replay_create() */
+		now = zpl_inode_get_timespec(vap->va_ctime);
```
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Can we make this timespec64 as well?

File: include/sys/zfs_znode.h [hunk 3]
```diff
@@ -291,19 +291,36 @@ zfs_inherit_projid(znode_t *dzp)
...
+ * Decode ZFS stored time values to a struct timespec
+ * 4.17 and older kernels.
+ */
 #define	ZFS_TIME_DECODE(tp, stmp)		\
 {						\
-	(tp)->tv_sec = (time_t)(stmp)[0];		\
-	(tp)->tv_nsec = (long)(stmp)[1];		\
+	(tp)->tv_sec = (time_t)(stmp)[0];	\
+	(tp)->tv_nsec = (long)(stmp)[1];	\
 }
```
Comment (tuxoko) [inline include/sys/zfs_znode.h]: We can take this opportunity to change these macro to `do { } while (0)`

**Comment by behlendorf** [comment]
> Can we make this timespec64 as well?

Yup, we can do better.  I've refreshed the PR and took a slightly different approach.  A new type was introduced called `inode_timespec_t` which can be used anywhere we need to a local copy of the times.

**Pull request merged**